src/Symfony/Component/Translation/Translator.php

```diff
@@ -72,17 +80,29 @@ class Translator implements TranslatorInterface, TranslatorBagInterface
     /**
      * Constructor.
      *
-     * @param string               $locale   The locale
-     * @param MessageSelector|null $selector The message selector for pluralization
-     * @param string|null          $cacheDir The directory to use for the cache
-     * @param bool                 $debug    Use cache in debug mode ?
+     * @param string                      $locale     The locale
+     * @param MessageFormatterInterface[] $formatters An array of formatters
+     *                                                Passing the MessageSelector as a second parameter is deprecated since version 3.1.
+     * @param string|null                 $cacheDir   The directory to use for the cache
+     * @param bool                        $debug      Use cache in debug mode ?
      *
      * @throws \InvalidArgumentException If a locale contains invalid characters
      */
-    public function __construct($locale, MessageSelector $selector = null, $cacheDir = null, $debug = false)
+    public function __construct($locale, $formatters = null, $cacheDir = null, $debug = false)
     {
         $this->setLocale($locale);
-        $this->selector = $selector ?: new MessageSelector();
+
+        if ($formatters instanceof MessageSelector) {
+            @trigger_error('Passing a MessageSelector instance into the '.__METHOD__.' as a second argument is deprecated since version 3.1 and will be removed in 4.0. Inject a MessageFormatterInterface[] instance instead.', E_USER_DEPRECATED);
+        }
+
+        if (null === $formatters || $formatters instanceof MessageSelector) {
```

Comment:
I would handle the case of ``$formatters instanceof MessageSelector`` in the previous if, at the same time than triggering the deprecation. It will be more readable (we can have ``$formatters = array(new TransMessageFormatter());`` in the case of ``null`` which will be easier to understand what is passed IMO) and will need that we have a single block to remove in 4.0, and it is the one containing the deprecation trigger